Issue 23009/separate angle and disable vsync

glowe opened the pull request:
The `--angle` and `--disable-vsync` options were declared as global options, but only used in the Glutin embedding for desktop builds. Moving them to the Glutin embedding code makes them easier to update in the future.

I modified `opts::from_cmdline_args` to accept a `getopts::Options` (as prescribed in the issue) and augmented `opts::ArgumentParsingResult` to include an `opts::Matches` and `content-process` String when appropriate. I could use some feedback on this last bit. I could have changed the function to return `opts::Matches` and have the embedding code look for the presence of `content-process`, but I felt that the approach I went with was closer to the original design.

The other aspect I'm not sure about is moving `disable-vsync` from a global debug option to a plain embedder option. This changes the command line interface for glutin, which is maybe bad. However I wasn't sure whether it was worth preserving the original behavior given the complexity of injecting debug options into `opts::from_cmdline_args`.

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes partially fix #23009 – there are 2 more options to deal with, but I'm not sure if we should handle them yet.
- [x] These changes do not require tests because this is a refactoring and I'm hoping that the existing tests cover these changes.

r? @jdm

Commit:
Remove some global options access within glutin

Removed opts::get() access for the two glutin specific options: angle
and disable-vsync. This is the first step in a refactoring to separate
these two options from the global options.

Commit:
Make angle and disable-vsync embedder options

The angle and disable-vsync options were declared as global options but
only used in the Glutin embedding for desktop builds. Moving them to
the Glutin embedding code makes them easier to update.

Partially fixes #23009

jdm commented:
I like these changes, and I appreciate the care you have taken in describing your reasoning for them! I think moving the disable-vsync flag to be a non-debug option is fine in this case. I've found several other flags that are glutin-only that we can convert as well: device_pixels_per_px, clean_shutdown, use_msaa, and no_native_titlebar.

Commit:
Extract 3 more embedder options

Extracted clean-shutdown, msaa, and no-native-titlebar embedder
specific options out from the global options.

Partially fixes #23009

glowe commented:
I extracted `clean_shutdown`, `use_msaa`, and `no_native_titlebar`. The `device_pixels_per_px` is a bit more difficult, because it's used in the Constellation pipeline:

https://github.com/servo/servo/blob/dc8be8f2825b7de873bf442aeb4a5dd0ed88d4b1/components/constellation/pipeline.rs#L612

jdm commented:
We can file a followup to remove the use from constellation by passing it as an argument to the constructor instead, if you'd like.

jdm commented on ports/libmlservo/Cargo.toml:
```diff
@@ -25,6 +25,7 @@ simpleservo = { path = "../libsimpleservo/api", features = ["no_static_freetype"
 rust-webvr = { version = "0.16", features = ["magicleap"] }
 webxr-api = { git = "https://github.com/servo/webxr", features = ["ipc"] }
 webxr = { git = "https://github.com/servo/webxr", features = ["ipc", "magicleap"] }
+getopts = "0.2.11"
```
Did this belong in libsimpleservo/api/Cargo.toml instead?

glowe replied:
Yes, you're right. Thanks for catching!

jdm reviewed (changes requested):
This looks great!

glowe commented:
> We can file a followup to remove the use from constellation by passing it as an argument to the constructor instead, if you'd like.

We can definitely make that change in this PR. However, I suspect that's going to force a modification of `servo::Servo::new`, which would be a public interface change. If so, does that mean we need to bump the version of `libservo`?

jdm commented:
libservo isn't published anywhere so we don't need to follow any semver rules.

jdm commented on ports/libsimpleservo/api/src/lib.rs:
```diff
@@ -159,6 +158,7 @@ pub fn init(
     gl: Rc<dyn gl::Gl>,
     waker: Box<dyn EventLoopWaker>,
     callbacks: Box<dyn HostTrait>,
+    device_pixels_per_px: Option<f32>,
```
This will require updating https://github.com/servo/servo/blob/6a54d600431621bfd4c4d4e49e7816e95deec68e/ports/libmlservo/src/lib.rs#L193 and https://github.com/servo/servo/blob/6a54d600431621bfd4c4d4e49e7816e95deec68e/ports/libsimpleservo/jniapi/src/lib.rs#L117 and https://github.com/servo/servo/blob/6a54d600431621bfd4c4d4e49e7816e95deec68e/ports/libsimpleservo/capi/src/lib.rs#L320.

However, I just noticed https://github.com/servo/servo/blob/6a54d600431621bfd4c4d4e49e7816e95deec68e/ports/libsimpleservo/api/src/lib.rs#L49 which is the value that is consulted when device_pixels_per_px is None in [this code](https://github.com/servo/servo/blob/master/components/compositing/compositor.rs#L1083-L1091). So I think we don't actually need to add this new argument; the device_pixels_per_px value is an override that is only meaningful in the glutin port, and the other ports already have a supported API for reporting the DPI factor, so we can revert this change to the `init` method.

glowe replied:
Thanks for catching this. I removed this change and pushed the rebased commit without this change.

jdm commented:
@bors-servo r+
Looks great! Thanks for tackling this work!

bors-servo commented:
:pushpin: Commit bf47d85 has been approved by `jdm`

<!-- @bors-servo r=jdm bf47d85640df64240edbee92f9e87ca1fb1441dc -->

glowe commented:
> Looks great! Thanks for tackling this work!

Thank you for your guidance. It was fun!

bors-servo commented:
:hourglass: Testing commit bf47d85640df64240edbee92f9e87ca1fb1441dc with merge 0b02cd35a845216d55be337ed942a242d306d4c7...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/R8BgRiC-TnCwA145W-1d4A)

glowe commented:
> 💔 Test failed - status-taskcluster

Looks like these errors are caused by me failing to update some things in `libsimpleservo`, however `mach build -d` built cleanly on my machine. I assume I didn't catch this issue, because `mach build -d` doesn't compile this crate? Is there an alternate invocation I can run that would catch these types of errors before submitting again?

CYBAI commented:
@glowe I guess you're using macOS? maybe it's easier to fix them and push to the PR and ask homu to try 🤔? 

```rust
error[E0433]: failed to resolve: use of undeclared type or module `Options`
   --> ports/libsimpleservo/api/src/lib.rs:173:33
    |
173 |         opts::from_cmdline_args(Options::new(), &args);
    |                                 ^^^^^^^ use of undeclared type or module `Options`

error[E0061]: this function takes 3 parameters but 2 parameters were supplied
   --> ports/libsimpleservo/api/src/lib.rs:207:17
    |
207 |     let servo = Servo::new(embedder_callbacks, window_callbacks.clone());
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 parameters

error: aborting due to 2 previous errors
```

jdm commented:
`mach build --libsimpleservo` is what you're looking for.

Commit:
Extract device_pixels_per_px from global opts

This is also an embedder specific option, so removing it from the
global options makes sense.

glowe commented:
Thanks for the help again. I believe I've addressed the libsimpleservo build errors.

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit 0ee3004 has been approved by `jdm`

<!-- @bors-servo r=jdm 0ee300479ec22e18ea00d9d1c7c60bdd35808983 -->

bors-servo commented:
:hourglass: Testing commit 0ee300479ec22e18ea00d9d1c7c60bdd35808983 with merge 2ad6e940913972620eab6c45a3fe2ace3c2c88f0...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13904), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13968), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/bT91GH_qSX6WXqfIpKxgIQ)
Approved by: jdm
Pushing 2ad6e940913972620eab6c45a3fe2ace3c2c88f0 to master...
